ukvm/ukvm_hv_openbsd.c

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright (c) 2015-2017 Contributors as noted in the AUTHORS file
+ *
+ * This file is part of ukvm, a unikernel monitor.
+ *
+ * Permission to use, copy, modify, and/or distribute this software
+ * for any purpose with or without fee is hereby granted, provided
+ * that the above copyright notice and this permission notice appear
+ * in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL
+ * WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE
+ * AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR
+ * CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS
+ * OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT,
+ * NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+ * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+/*
+ * ukvm_hv_openbsd.c: Architecture-independent part of OpenBSD vmm(4) backend
+ * implementation.
+ */
+
+#include <assert.h>
+#include <err.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#define _WITH_DPRINTF
+#include <stdio.h>
+
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+#include <sys/types.h>
+#include <sys/sysctl.h>
+#include <machine/vmmvar.h>
+#include <sys/param.h>
+
+#include "ukvm.h"
+#include "ukvm_hv_openbsd.h"
+
+/*
+ * TODO: To ensure that the VM is correctly destroyed on shutdown (normal or
+ * not) we currently install an atexit() handler. The top-level API will need
+ * to be changed to accomodate this, e.g. by introducing a ukvm_hv_shutdown(),
+ * however this is incompatible with the current "fail fast" approach to
+ * internal error handling.
+ */
+static struct ukvm_hv *cleanup_hv;
+
+static void cleanup_vmd_fd(void)
+{
+    if (cleanup_hv != NULL && cleanup_hv->b->vmd_fd != -1)
+        close(cleanup_hv->b->vmd_fd);
+}
+
+static void cleanup_vm(void)
+{
+    if (cleanup_hv != NULL && cleanup_hv->b->vcp_id != -1) {
+        struct vm_terminate_params vtp = { .vtp_vm_id = cleanup_hv->b->vcp_id };
+        if (ioctl(cleanup_hv->b->vmd_fd, VMM_IOC_TERM, &vtp) < 0)
+            err(errno, "terminate vmm ioctl failed - still exiting");
```

Comment:
I think our tests are checking for return code 1 in case of any error. Also, err() will print the errno. You could change this to just `err(1, "ukvm...")`